### ISSUE TITLE
memory usage of `player` class

### ISSUE THREAD

**Issue by lolanchen** [hsahovic/poke-env#164]
Hi Harris how are you doing!

TL;DR: 
1. the player class seems to be using to much memory, how do I stop it from doing so?
2. cool down time for between games for the `Player` class 
----------------------

I'm currently using a custom `train_on_ladder` function, which simply replace the send/receive challenge part for `launch_battle` with laddering in `play_against`,   
to train 16 agents against each other using the ladder in a self-play-ish setup.  (code attached below)

To add some variety in the training pool, I also launched 4 `MaxDamagePlayer` to the ladder using a small script I wrote(code attached below).   
I used your implementation of `MaxDamagePlayer` in [this example](https://github.com/hsahovic/poke-env/blob/master/examples/max_damage_player.py), which inherits the `Player` class.

After a million train steps, I found that the 4 `MaxDamagePlayer` combined are taking up about 30GB of memory, and the size would keep growing as they play more battles. And this is not good.   
I suspect maybe the `Player` class is storing some weird things in the memory, but couldn't figure it out.  
Does the `Player` class store anything? if so, how can I prevent it from doing it? 

Also unlike `env_player` which has a `PAUSE_BETWEEN_RETRIES` attribute,
`Player` does not have any cool down time, which in my setup would cause a lot of wasted computation (half of the games would be played between the MaxDamagePlayers, no agent involved).  
It would be nice to have a cool down time for the `Player` class too.

Cheers!  

---------------------

train on ladder
```python
# File: player/env_player.py
    def train_on_ladder(
        self, env_algorithm: Callable, env_algorithm_kwargs=None
    ):
        self._start_new_battle = True

        async def launch_battles(player: EnvPlayer):
            battles_coroutine = asyncio.gather(
                player.ladder(n_games=1)
            )
            await battles_coroutine

        def env_algorithm_wrapper(player, kwargs):
            env_algorithm(player, **kwargs)
            player._start_new_battle = False
            while True:
                try:
                    player.complete_current_battle()
                    player.reset()
                except OSError:
                    break

        loop = asyncio.get_event_loop()

        if env_algorithm_kwargs is None:
            env_algorithm_kwargs = {}

        thread = Thread(
            target=lambda: env_algorithm_wrapper(self, env_algorithm_kwargs)
        )
        thread.start()

        while self._start_new_battle:
            loop.run_until_complete(launch_battles(self))
        thread.join()
```

launch_max.py
```python
from opponents import MaxDamagePlayer
from poke_env.player_configuration import PlayerConfiguration
import time
import os
import asyncio
from threading import Thread

# name the player with it's pid
def indexgen_pid(digits:int) -> str:
    return format(os.getpid,f"0{digits}d")

async def main():
    maxie = MaxDamagePlayer(battle_format='gen8randombattle', player_configuration=PlayerConfiguration('Max'+indexgen_pid(5), None))

    # enough games until the training is over
    await maxie.ladder(n_games=10**10)

if __name__ == "__main__":
    asyncio.get_event_loop().run_until_complete(main())
```

**Comment by hsahovic** [comment]
Hey @lolanchen,

Thanks for bringing this up.

### On RAM usage

`Player` objects store `Battle` objects, so a linear increase in RAM usage is expected over time if the battle store is not reset. I wouldn't expect RAM usage to go so high, but I can see something like that happening over > 10k battles per player, since some rather large json messages are stored by each `Battle` object.

There are two ways to fix this:

1. Call [`player.reset_battles()`](https://poke-env.readthedocs.io/en/stable/player.html?highlight=reset%20battles#poke_env.player.player.Player.reset_battles) every x battles. This should reset the player's battle store and free up the ram if the `Battle` objects are not referenced elsewhere
2. `del player` from time to time and re-add new players

### On introducing delays

I think the easiest way to achieve this is to do something like this:

```python
class MaxDamagePlayer(Player)
    ...
    async def ladder(self, n_games):
        await sleep(DELAY)
        await super().ladder(n_games)  # This might need a tweak, I haven't tested it
````

**Comment by lolanchen** [comment]
@hsahovic 
Thanks as always!   

for the delay, I ended up just adding the sleep in the ladder funciton in `player.py` and it works fine.   

for the memory issue, calling `player.reset_battles()` does seem to solve the problem,   
but I thought it'd be nice if `Player` class can offer an option to save/not save battles on initialization. 
something like 
```python
p0 = Player() # saves battles by default
p1 = Player(save_battles=False) # won't save battles
```

cheers

**Comment by hsahovic** [comment]
That's a good idea. This would have implications on other things (such as `Player.win_rate`), so I need to think about it a bit more, but maybe having a limit to the number of stored battles would be a good solution.

**Comment by lolanchen** [comment]
@hsahovic 
 On the same RAM usage issue, it seems that resetting battles only solves half of the problem. The memory usage still increase linearly, just at a slightly slower rate.

  some quick debugging with `tracemalloc` implies the problem is in the [listen function](https://github.com/hsahovic/poke-env/blob/master/src/poke_env/player/player_network_interface.py#L243) in `player_network_interface.py`  
This `listen()` function has a try-except-finally block and looks like the `final` block (not sure what `coroutine.cancel()` does but I guess it releases memory?)is never executed.   
I checked by putting `print(len(coroutines))` after line 254 and the number just keep growing.   
I suspect this is eating extra memory too.

  I was training 36 agents against each other on the ladder, the initial RAM usage was around 14GB and at 1M step it went up to 100GB. This is with all player's battle buffer cleared with `reset_battle()`

Please look into this.

**Comment by noojHeron** [comment]
@hsahovic 

Hi,

I am running into a similar issue when training agents. I have tried both the `reset_battles()` solution as well as the `del player` solution to no avail. Similarly to lolanchen, it seems that coroutines are just piling up, and `coroutines.cancel()` is never being called. I'm a little confused as to why simply deleting the player classes after they are used does not seem to solve this issue. I am currently debugging the `try` statement in `player_network_interface.py`'s `listen()` and will post anything that might be useful to others.

Currently, it seems that nothing ever returns from the `try`, and neither the `finally` nor any of the exceptions are reached. This is after 100,000 timesteps and over a hundred games.

**Comment by noojHeron** [comment]
@hsahovic 

Hi,

Just wanted to post an update to the memory leak error. According to [this post on StackOverflow](https://stackoverflow.com/questions/49262379/does-finally-always-execute-in-python), a `finally` block in an asynchronous function may never execute. It seems that `try-finally` is not a safe operation for asynchronous processes.

Because this is unintended behavior that prevents training over long periods of time, I would like to request that this thread be marked as a `BUG`. Thanks!

**Comment by hsahovic** [comment]
@noojHeron Good catch!

Coroutines being stored is removed in [this commit](https://github.com/hsahovic/poke-env/commit/c53a29341554be6627a6c6414fbe4fc1f235dc8b); it will be released in `0.4.17`.
